assets/js/components/Dashboard/Dashboard.tsx

```diff
@@ -6,45 +6,134 @@ import { Alert } from "../../models/alert";
-}
 type ContextType = {
   places: Place[];
   alerts: Alert[];
   screensByAlertMap: ScreensByAlert;
 };
 
+interface BannerAlert {
+  alert?: Alert;
+  closedAt?: string;
+}
+
 const Dashboard: ComponentType = () => {
   const [places, setPlaces] = useState<Place[]>([]);
   const [alerts, setAlerts] = useState<Alert[]>([]);
   const [screensByAlertMap, setScreensByAlertMap] = useState<ScreensByAlert>(
     {}
   );
+  const [bannerAlert, setBannerAlert] = useState<BannerAlert>();
 
   useEffect(() => {
-    fetch("/api/alerts")
-      .then((response) => response.json())
-      .then(({ alerts, screens_by_alert }: AlertsResponse) => {
-        setAlerts(alerts);
-        setScreensByAlertMap(screens_by_alert);
+    fetchPlaces().then((placesList) => setPlaces(placesList));
+
+    fetchAlerts().then(({ alerts: newAlerts, screens_by_alert }) => {
+      setAlerts((oldAlerts) => {
+        findAndSetBannerAlert(oldAlerts, newAlerts);
+        return newAlerts;
       });
+      setScreensByAlertMap(screens_by_alert);
+    });
   }, []);
 
-  useEffect(() => {
-    fetch("/api/dashboard")
-      .then((response) => response.json())
-      .then((placeList: []) => {
-        setPlaces(placeList);
+  // Fetch alerts every 4 seconds.
+  useInterval(() => {
+    fetchAlerts().then(({ alerts: newAlerts, screens_by_alert }) => {
+      setAlerts((oldAlerts) => {
+        findAndSetBannerAlert(oldAlerts, newAlerts);
+        return newAlerts;
       });
-  }, []);
+      setScreensByAlertMap(screens_by_alert);
+    });
+  }, 4000);
+
+  const findAndSetBannerAlert = (oldAlerts: Alert[], newAlerts: Alert[]) => {
+    const now = moment().utc().format();
+    const closedAlert = getFirstClosedAlert(oldAlerts, newAlerts);
+    const postedOrEditedAlert = getPostedOrEditedAlert(newAlerts);
+    // If there is a closed alert, just show it and save when it was closed.
+    if (closedAlert) {
+      setBannerAlert({ alert: closedAlert, closedAt: now });
+    }
+    // If there is not a new closed alert but
+    // the current bannerAlert does not have a defined closedAt
+    // (meaning there isn't a current banner alert or it is a posted/edited alert)
+    else if (!bannerAlert?.closedAt) {
+      setBannerAlert({ alert: postedOrEditedAlert });
+    }
+    // If the current bannerAlert has a closedAt but a recent posted/edited alert came after the closed bannerAlert
+    else if (
+      postedOrEditedAlert &&
+      moment(postedOrEditedAlert.created_at).isAfter(
+        moment(bannerAlert.closedAt) ||
+          moment(postedOrEditedAlert.updated_at).isAfter(
+            moment(bannerAlert.closedAt)
+          )
+      )
+    ) {
+      setBannerAlert({ alert: postedOrEditedAlert });
+    }
+    // If no other alert is eligible to be the bannerAlert, expire the current bannerAlert if two minutes have passed since it started displaying.
+    else if (
+      moment(now).isSameOrAfter(moment(bannerAlert.closedAt).add(2, "minutes"))
```

Comment:
Would this check handle posted/edited alerts? Since those alerts wouldn't have a defined `closedAt` value according to the comment on lines 65-67

Reply:
That scenario should not get through the first `else if` on line 68. If `bannerAlert` is `undefined` or `bannerAlert.closedAt` is `undefined` and there is no new closed alert, it will set the `bannerAlert` to the result of `getPostedOrEditedAlert`. That result can either be a posted/edited alert from the last two minutes or `undefined` if there is no recent alert. 

After not looking at this for a week, this conditional block is hard to follow, but I'm really not sure how to simplify it 😬 . If you have any suggestions, let me know.

Reply:
Hmm yeah I've been looking at it and can't think of any ways to make any significant simplifications. One small suggestion that might help is putting the `moment` comparisons on Lines 74-77 in a separate helper function to help the conditional block read a bit better, but I think that's the best I can come up with at the `moment` 🤷 

One thing I'm a little confused on is how a posted/edited alert banner will be closed after two minutes. I see that for a closed alert, we're constantly checking if two minutes has passed since the `closedAt` time, but a posted or edited alert wouldn't have that field, right? Oh, or does that check implicitly happen in `getPostedOrEditedAlert` and then the banner alert would just get set as `undefined` on line 69?

Reply:
Yep, it happens in `getPostedOrEditedAlert`. For recently posted/edited alerts, we are able to check if `created_at` or `updated_at` is in the last two minutes. Because the alert itself has that data, we don't need to store it and constantly check against the time we first saw it. For a closed alert, those fields are unreliable and would be lost on the second refresh after the alert closed anyway.